### PR TITLE
Complimentary gluon DataLoader improvements

### PR DESCRIPTION
**Pull request by zhreshold**
## Description ##
This PR provides complimentary improvements for user to customize DataLoader in various use cases for optimal performance.
 
- Now DataLoader supports `ThreadPool` version when shared memory is not prefered, or GIL is not a problem. Ref discussion in #13447
- Support advanced configurations of threading management. Multiprocessing workers can use custom number of threads to process data rather than limited to 1. #13593 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: python/mxnet/gluon/data/dataloader.py [hunk 1]
```diff
@@ -551,14 +568,17 @@ def same_process_iter():
...
 
     def __len__(self):
         return len(self._batch_sampler)
 
     def __del__(self):
         if self._worker_pool:
-            # manually terminate due to a bug that pool is not automatically terminated on linux
+            # manually terminate due to a bug that pool is not automatically terminated
+            # https://bugs.python.org/issue34172
             assert isinstance(self._worker_pool, multiprocessing.pool.Pool)
```
Comment (kohillyang) [inline python/mxnet/gluon/data/dataloader.py]: What happens here if threadpool is True?

File: src/initialize.cc [hunk 2]
```diff
@@ -57,13 +57,18 @@ class LibraryInitializer {
         Engine::Get()->Start();
       },
       []() {
-        // Make children single threaded since they are typically workers
-        dmlc::SetEnv("MXNET_CPU_WORKER_NTHREADS", 1);
-        dmlc::SetEnv("OMP_NUM_THREADS", 1);
+        // Conservative thread management for multiprocess workers
+        const size_t mp_worker_threads = dmlc::GetEnv("MXNET_MP_WORKER_NTHREADS", 1);
+        const size_t mp_omp_threads = dmlc::GetEnv("MXNET_MP_OMP_NUM_THREADS", 1);
```
Comment (anirudh2290) [inline src/initialize.cc]: libgomp doesnt play well with fork: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52303 . Which means user using libgomp and increasing the num threads can run into a deadlock.

Comment (zhreshold) [reply src/initialize.cc]: I see, do you suggest completely turn off omp in worker process regardless?

**Comment by YutingZhang** [comment]
@zhreshold Thank you! When is it going to be merged?

**Comment by zhreshold** [comment]
@szha @eric-haibin-lin for review/merge

**Comment by YutingZhang** [comment]
Thanks!

**Comment by YutingZhang** [comment]
Is it include the `mxnet-cu90mkl --pre` now?

**Comment by leezu** [comment]
@YutingZhang you can check the latest published file here https://pypi.org/project/mxnet-cu90mkl/#history If the build date is after the merge date, it is included

**Comment by YutingZhang** [comment]
@leezu Yes. I had been using it over the weekend :)

**Pull request merged**